From 9e6b69aef8b7088f634eac47318ea77031b10747 Mon Sep 17 00:00:00 2001
From: Yongli He <yongli.he@windriver.com>
Date: Thu, 4 Sep 2008 14:57:37 +0800
Subject: [PATCH] raid1 instrumention config option and misc fixes

Port/fixes for raid1_bio_instrumentation.patch and
raid1_mirror_index_instrumentation.patch.

Signed-off-by: Amy Fong <amy.fong@windriver.com>
Integrated-by: Yongli he   <yongli.he@windriver.com>
---
 drivers/md/Kconfig      |    8 ++++++++
 drivers/md/md.c         |    4 ++--
 drivers/md/raid1.c      |   10 +++++++---
 include/linux/raid/md.h |    4 ++++
 4 files changed, 21 insertions(+), 5 deletions(-)

diff --git a/drivers/md/Kconfig b/drivers/md/Kconfig
index 07d92c1..80d75c2 100644
--- a/drivers/md/Kconfig
+++ b/drivers/md/Kconfig
@@ -85,6 +85,14 @@ config MD_RAID1
 
 	  If unsure, say Y.
 
+config MD_RAID1_INSTRUMENTATION
+	bool "Support for RAID1 instrumentation"
+	depends on MD_RAID1
+	---help---
+	  Enables instrumentation for RAID1
+
+	  If unsure, say N.
+
 config MD_RAID10
 	tristate "RAID-10 (mirrored striping) mode (EXPERIMENTAL)"
 	depends on BLK_DEV_MD && EXPERIMENTAL
diff --git a/drivers/md/md.c b/drivers/md/md.c
index 52d65c2..0be476e 100644
--- a/drivers/md/md.c
+++ b/drivers/md/md.c
@@ -72,7 +72,7 @@ static void autostart_arrays (int part);
 static LIST_HEAD(pers_list);
 static DEFINE_SPINLOCK(pers_lock);
 
-static void md_print_devices(void);
+void md_print_devices(void);
 
 static DECLARE_WAIT_QUEUE_HEAD(resync_wait);
 
@@ -1648,7 +1648,7 @@ static void print_rdev(mdk_rdev_t *rdev)
 		printk(KERN_INFO "md: no rdev superblock!\n");
 }
 
-static void md_print_devices(void)
+void md_print_devices(void)
 {
 	struct list_head *tmp, *tmp2;
 	mdk_rdev_t *rdev;
diff --git a/drivers/md/raid1.c b/drivers/md/raid1.c
index d1b47dc..2f733b7 100644
--- a/drivers/md/raid1.c
+++ b/drivers/md/raid1.c
@@ -255,7 +255,7 @@ static inline void update_head_pos(int disk, r1bio_t *r1_bio)
 		r1_bio->sector + (r1_bio->sectors);
 }
 
-#if RAID1_INSTRUMENTATION
+#ifdef CONFIG_MD_RAID1_INSTRUMENTATION
 
 #define indent_printk(indentation_level, format_string, ...) \
 	printk(KERN_DEBUG "%*s" format_string, indent_level, " ", ##__VA_ARGS__)
@@ -410,7 +410,7 @@ static void raid1_end_write_request(struct bio *bio, int error)
 		if (r1_bio->bios[mirror] == bio)
 			break;
 
-#if RAID1_INSTRUMENTATION
+#ifdef CONFIG_MD_RAID1_INSTRUMENTATION
 	/*
 	 * There's a chance that the above loop will not find a matching index
 	 * of the mirror associated with the passed in struct bio *bio. We
@@ -419,7 +419,9 @@ static void raid1_end_write_request(struct bio *bio, int error)
 	 */
 	if (unlikely(mirror >= conf->raid_disks)) {
 		int cpu = get_cpu();
+#ifdef CONFIG_PANIC_LOGS
 		kcore_logon(0);
+#endif
 		printk(KERN_DEBUG "raid1.c: (CPU %d) None of the bio's in the "
 				"bios array matched the passed in bio struct. "
 				"[mirror=%d,conf->raid_disks=%d]\n", cpu,
@@ -429,12 +431,14 @@ static void raid1_end_write_request(struct bio *bio, int error)
 		printk(KERN_DEBUG "contents of passed in bio:\n");
 		dump_bio(bio);
 		printk(KERN_DEBUG "partial contents of r1_bio:\n");
-		dump_r1bio_t(r1_bio, raid_disks);
+		dump_r1bio_t(r1_bio, conf->raid_disks);
 		printk(KERN_DEBUG "partial contents of r1_bio->mddev:\n");
 		dump_mddev_t(r1_bio->mddev);
 
 		put_cpu();
+#ifdef CONFIG_PANIC_LOGS
 		kcore_logoff();
+#endif
 	}
 #endif
 
diff --git a/include/linux/raid/md.h b/include/linux/raid/md.h
index dc0e3fc..596833b 100644
--- a/include/linux/raid/md.h
+++ b/include/linux/raid/md.h
@@ -98,6 +98,10 @@ extern void md_new_event(mddev_t *mddev);
 extern int md_allow_write(mddev_t *mddev);
 extern void md_wait_for_blocked_rdev(mdk_rdev_t *rdev, mddev_t *mddev);
 
+extern void md_print_devices (void);
+
+#define MD_BUG(x...) { printk("md: bug in file %s, line %d\n", __FILE__, __LINE__); md_print_devices(); }
+
 #endif /* CONFIG_MD */
 #endif 
 
-- 
1.5.5.1

